### ISSUE TITLE
setting `doom-unicode-font` removes fallback fonts that support emojis

### ISSUE THREAD

**Issue by vancleve** [doomemacs/doomemacs#7036]
### I confirm that...

- [X] I have searched the [issue tracker](https://github.com/doomemacs/doomemacs/issues), [documentation](https://docs.doomemacs.org), [FAQ](https://docs.doomemacs.org/-/faq), [Discourse](https://discourse.doomemacs.org), and [Google](https://google.com), in case this issue has already been reported/resolved.

- [X] I have read "[How to Debug Issues](https://doomemacs.org/d/how2debug)", and will use it to provide as much information about this issue as possible.

- [X] The issue can be reproduced on the **latest** available commit of Doom.

- [X] The issue can be reproduced on a stable release of Emacs, such as 27 or 28. *(Doom does not support development builds like 29+ or any version ending in .50 or .9x)*

### Expected behavior

When I set `doom-unicode-font` I expect emoji support to continue working.

### Current behavior

When I set `doom-unicode-font`, my emoji support disappears. Maybe because of this code here?

https://github.com/doomemacs/doomemacs/blob/e96624926d724aff98e862221422cd7124a99c19/lisp/doom-ui.el#L554

### Steps to reproduce

1. Set `(setq doom-unicode-font (font-spec :family "TeX Gyre Termes Math"))` in `config.el`
2. Try `M-x emojify-insert-emoji`
3. Most emojis missing whereas before there were present before setting `doom-unicode-font`

### System Information

https://pastebin.com/p2M1KnAH

**Comment by M1ndo** [comment]
@vancleve Hey, Tex Gyre Termes doesn't have emoji supports, only math glyphs and some unicode characters support.
The emoji fallback font families are described as
```lisp
(defvar doom-emoji-fallback-font-families
  '("Apple Color Emoji"
    "Segoe UI Emoji"
    "Noto Color Emoji"
    "Noto Emoji")
```
You can customize and set the fallback-font-families or use another unicode-font like [JuliaMono](https://juliamono.netlify.app/) which will work very well for both Latex in (LuaTex or Xelatex) and Emacs.

**Comment by vancleve** [comment]
Hi @M1ndo! I haven't changed `doom-emoji-fallback-font-families` and it retains the font families you show above which work when I don't set `doom-unicode-font`.

The problem as far as I can tell is that here,
https://github.com/doomemacs/doomemacs/blob/e96624926d724aff98e862221422cd7124a99c19/lisp/doom-ui.el#L554,
the code discards `doom-emoji-fallback-font-families` when you set `doom-unicode-font` instead of adding those families to the fontset `unicode` after the `doom-unicode-font`.

**Comment by M1ndo** [comment]
Hey @vancleve Sorry for the delay. Yes you are absolutely right, `doom-emoji-fallback-font-families` is a fallback for when `doom-unicode-font` is not set, Consider when setting `doom-unicode-font` to use a `unicode-font` that's has enough unicode support like the ones i suggested.
If that's convenient for you but still want to use `TeX Gyre Termes Math` In LaTeX Mode or Any Major mode, you could use mixed-pitch-mode with a custom font like `TeX Gyre`, you could view how to achieve this from my mixed-pitch [config](https://github.com/M1ndo/dotfiles/blob/master/.doom.d/config.org#mixed-pitch)

**Comment by vancleve** [comment]
Thanks for the example code, I'll take a look!

I guess my issue here though, and why I think this could a "bug" or feature request at the least, is that I don't see why one can't keep the `doom-emoji-fallback-font-families` in the font-set while also setting `doom-unicode-font` to `TeX Gyre Termes Math` or whatever one chooses. 

If I use `describe-fontset` in AUCTeX, I get something like this
<img width="425" alt="image" src="https://user-images.githubusercontent.com/1442334/215225289-a6a2e160-a8b3-44c5-931c-ed51f90545f1.png">
and I'm thinking maybe all of these fonts are fallbacks for this unicode range? So my question is why can't we keep the `doom-emoji-fallback-font-families` in this list too?

**Comment by M1ndo** [comment]
@vancleve As much as i know this is not a bug, but could be a feature request!. `doom-unicode-font` is a primary fallback replacement (when set) for `doom-emoji-fallback-font-families` and vice versa.
what i mean, is that if you append both `doom-unicode-font` and `doom-emoji-fallback-font-families` they can be conflicted with each other, if one glyphs exist on both one can take over the other and bad things happen but we don't want that, we want when a glyph is missing the fallback can kick in and this is pretty much not possible if a glyph exists on both. You'd have to disable one font or disable the glyph from the font you chose or make sure its missing on one of them.

Disabling a glyph on two fonts can be a hassle if both fonts have hundreds/thousands of glyphs, in your case `TeX Gyre Termes Math` does have only few, so having a fallback will only work for this case.
You can enable these modules in ui `(emoji +unicode)`  `unicode` which will fix the missing emojis by using `emojify-mode` and `unicode-fonts`.
If some are still missing you can append `doom-emoji-fallback-font-families` if you want.
```lisp
(when (fboundp 'set-fontset-font)
  (dolist (font (list doom-emoji-fallback-font-families))
    (set-fontset-font t 'unicode font nil 'append)))
```

**Comment by vancleve** [comment]
Thanks so much for the help @M1ndo!

This helps me get a much better handle on what emacs is doing font-wise. Before modifying `doom-unicode-font`, what I was seeing in AUCTeX mode was that for a collection of unicode characters, say $\mathcal{A} \mathcal{B} \mathcal{C}\ldots \mathcal{Z}$, some of the symbols were present, some where missing, 
<img width="472" alt="image" src="https://user-images.githubusercontent.com/1442334/215368624-65d63e9e-4da1-4517-b861-730e1c8d7525.png">
and the ones that were there were coming from at least two different fonts:
<img width="637" alt="image" src="https://user-images.githubusercontent.com/1442334/215368654-441e9871-bb08-4fa3-a400-66fcdbed25a4.png">
<img width="582" alt="image" src="https://user-images.githubusercontent.com/1442334/215368741-5eca84ce-6dc6-4a30-a64a-71893d223715.png">
<img width="583" alt="image" src="https://user-images.githubusercontent.com/1442334/215368793-b7143c7b-b6e9-4c1c-aa03-ac1c90845cf6.png">

Even more weirdly, I can't find `Arial Unicode MS` using `describe-fontset`, so I don't know how emacs is choosing that font!

**Comment by vancleve** [comment]
Adding this to my `config.el` seems to work:

```
(defun add-back-emoji-fallback-font-families ()
  (when (fboundp 'set-fontset-font)
    (let ((fn (doom-rpartial #'member (font-family-list))))
      (when-let (font (cl-find-if fn doom-emoji-fallback-font-families))
        (set-fontset-font t 'unicode font nil 'append)))))

(add-hook 'after-setting-font-hook 'add-back-emoji-fallback-font-families)
```

**Comment by M1ndo** [comment]
@vancleve No problem, I'm currently looking into this issue further and got some unexpected things. I'm not really sure if its from emacs 30 side or my configuration (fonts i'm using). Can you provide me with your full doom font setup?

**Comment by vancleve** [comment]
Sure!

```lisp
(setq doom-font (font-spec :family "FiraCode Nerd Font" :size 12)
      doom-variable-pitch-font (font-spec :family "Fira Sans")
      doom-unicode-font (font-spec :family "TeX Gyre Termes Math")
      )

;; add back the emoj fallback font families that aren't used when setting doom-unicode-font
(defun add-back-emoji-fallback-font-families ()
  (when (fboundp 'set-fontset-font)
    (let ((fn (doom-rpartial #'member (font-family-list))))
      (when-let (font (cl-find-if fn doom-emoji-fallback-font-families))
        (set-fontset-font t 'unicode font nil 'append)))))

(add-hook 'after-setting-font-hook 'add-back-emoji-fallback-font-families)
```

**Comment by M1ndo** [comment]
@vancleve Thanks!. I have further investigated the issue and i was very surprised to see the results.
To foreshadow i'm currently running `emacs 30.0.50`, i've switched up to your fontset configuration and i've done some testing with it. Most of the issues you've had could not be reproduced in this setup, except for "Emojis" (Without :ui (emoji) module or disabling emojify-mode).

![image](https://user-images.githubusercontent.com/44820142/215881227-574f42de-868d-4cdc-bb7e-219c420dc6ac.png)

In this screenshot you can see its exactly your font setup, except not adding  `fallback-unicode-fonts` and  `unicode` module is disabled.
Still all math glyphs, including non-ascii characters, multiple languages are all loaded perfectly fine, which is very surprising to me. In fact multiple fonts were being used simultaneously even without `unicode` module which uses [Unicode-Fonts](https://github.com/rolandwalker/unicode-fonts) Emacs does this out of the box tho configuring font mappings via fontsets is hard, Perhaps this was resolved on emacs `29-30`

![image](https://user-images.githubusercontent.com/44820142/215763192-5e7ae26a-aa41-479d-9bd4-3792d0a50aa3.png)
This is my setup with `unicode` module enabled, its much better than without it and emojis are displayed perfectly even without (emojify-mode).

In conclusion its pretty recommended if you enable `unicode` module, it sure will resolve any future issues dealing with any what we've saw. I'm also certain you can remove that piece of code you've added to append fallback fonts once you've done that everthing should work perfectly fine (I hope).

**Comment by vancleve** [comment]
Cool. It seems like the unicode module is definitely helping and replaces the emoji fallback.

I like having the same emoji font in emacs as in the rest of macOS so I'm keeping the fallback for that reason :-)

**Comment by vincentbernat** [comment]
Since the switch to Nerd Font, `doom-unicode-font` is always defined and therefore, the fallback to emoji fonts doesn't work anymore.

**Comment by hlissner** [comment]
@vincentbernat Adding `(setq doom-unicode-font nil)` to `$DOOMDIR/config.el` would undo. Or setting it to your preferred unicode font.

That said, does adding `"Symbols Nerd Font"` to `doom-symbol-fallback-font-families` (or `doom-emoji-fallback-font-families`) fix this issue? E.g.

```elisp
(add-to-list 'doom-symbol-fallback-font-families "Symbols Nerd Font")
```

**Comment by vincentbernat** [comment]
Yes, it works for me.

**Comment by agzam** [comment]
I've been trying to figure out why all emoji's are broken now on Mac. Coincidentally, I just had to install everything anew on a clean machine and wasted two hours trying "to fix my fonts" until I realized they are broken due to recent changes in Doom.

Like `C-x 8 RET` menu doesn't display any of them. Aren't they supported natively? Maybe the new changes require updating something in the user config? Why inserting a symbol in `emacs -Q`, displays them, but not in Doom?

**Comment by vancleve** [comment]
@hlissner, the above issues are why I think the default behavior should be to append `doom-emoji-fallback-font-families` to the unicode range in `‎set-fontset-font‎` if `doom-unicode-font` is set range rather than only using `doom-unicode-font` if its set.

**Comment by vancleve** [comment]
This is fixed by #7448
